### PR TITLE
Add `qml.WireCut` operator and initial `qcut` package

### PR DESCRIPTION
**Pull request by anthayes92**
**Context:**
This is the first addition to the circuit cutting compiler where we initially want a user to be able to manually place a wire cut during creation of a `qnode`.

**Description of the Change:**
Add a `WireCut` operator to be used during creation of a qnode to specify cut locations
Add initial `qcut` package

**Benefits:**
Will allow for intuitive placement of wire cuts

**Possible Drawbacks:**
Not a traditional quantum operator; this is a placeholder for the actual under-the-hood operations. Potential for confusion.

**Related GitHub Issues:**

### CODE REVIEW

File: tests/ops/qubit/test_non_parametric_ops.py [hunk 1]
```diff
@@ -547,6 +547,29 @@ def barrier():
...
+            qml.PauliX(wires=0)
+            qml.WireCut(wires=0)
+            return qml.state()
+
+        @qml.qnode(dev)
+        def without_wirecut():
+            qml.PauliX(wires=0)
+            return qml.state()
+
+        assert with_wirecut().all() == without_wirecut().all()
```
Comment (trbromley) [inline tests/ops/qubit/test_non_parametric_ops.py:570]: What does this line do? Maybe it works, although I was expecting to see an `np.allclose()` :thinking:

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 2]
```diff
@@ -1189,3 +1189,24 @@ def label(self, decimals=None):
 
     def adjoint(self):
         return Barrier(wires=self.wires)
+
+
+class WireCut(Operation):
+    r"""WireCut(wires)
+    The wire cut operator, used to manually place qubit cut locations.
```
Comment (trbromley) [inline pennylane/ops/qubit/non_parametric_ops.py:1196]:
```suggestion
    The wire cut operator, used to manually mark locations for wire cuts.
```
Might not matter too much, but wanted to emphasize that this operation is more of a marker than actually placing the cut.

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 3]
```diff
@@ -1189,3 +1189,24 @@ def label(self, decimals=None):
...
+
+    * Number of wires: AnyWires
+    * Number of parameters: 0
+
+    Args:
+        wires (Sequence[int] or int): the wires the operation acts on
+    """
+    num_wires = AnyWires
+    grad_method = None
+
```
Comment (trbromley) [inline pennylane/ops/qubit/non_parametric_ops.py:1208]: We need an `adjoint()` method too

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @anthayes92, looks great! We're missing:
- An `adjoint()` method
- Accommodation for circuit drawing - similar to `Barrier`, it'd be nice for text and MPL based drawing to not directly use the `WireCut` word but a more accurate symbol.

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 4]
```diff
@@ -1189,3 +1189,24 @@ def label(self, decimals=None):
...
+    def expand(self):
+        with qml.tape.QuantumTape() as tape:
+            ...
+        return tape
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/ops/qubit/non_parametric_ops.py:1212]: Yep, I recommend using `decompose()` here (like the operations do), and not `expand` :) It is more future compatible with the changes coming down the pipeline.

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -4,6 +4,11 @@
 
 <h3>New features since last release</h3>
 
+* Development of a `qcut` package has begun to allow for a circuit cutting
+  compiler and a `WireCut` operator has been added for manual wire cut placement
+  when constructing a QNode.
+  [(#2093)](https://github.com/PennyLaneAI/pennylane/pull/2093)
+
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply doc/releases/changelog-dev.md:11]: I think this makes sense for now, happy to adapt as it grows

File: doc/code/qml_qcut.rst [hunk 6]
```diff
@@ -0,0 +1,9 @@
+qml.qcut
+===============
+
+.. currentmodule:: pennylane.qcut
+
+.. automodapi:: pennylane.qcut
+    :no-heading:
+    :include-all-objects:
+    :skip:
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply doc/code/qml_qcut.rst:9]: Cool, makes sense to remove from this PR and working in the `transforms` module sounds good. Where do you envision the graph partitioning will live? There is also a fair few potential additions to the circuit cutter, perhaps a `qcut_utils.py` file would come in handy further down the line

File: tests/ops/qubit/test_non_parametric_ops.py (hunk 1 above)
Comment (anthayes92) [reply tests/ops/qubit/test_non_parametric_ops.py:570]: updated thanks!

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 7]
```diff
@@ -1189,3 +1189,30 @@ def label(self, decimals=None):
...
+    Args:
+        wires (Sequence[int] or int): the wires the operation acts on
+    """
+    num_wires = AnyWires
+    grad_method = None
+
+    def decompose(self):
+        with qml.tape.QuantumTape() as tape:
+            ...
+        return tape
```
Comment (trbromley) [inline pennylane/ops/qubit/non_parametric_ops.py:1212]:
```suggestion
    # pylint: disable=unused-argument
    def decomposition(self, wires):
        return []
```
I think [decompose](https://github.com/PennyLaneAI/pennylane/blob/b3895ea6b84946b3e2cbca6bcc8ef45566ab661e/pennylane/operation.py#L550) is not intended to be overridden.

File: tests/transforms/test_adjoint.py [hunk 8]
```diff
@@ -65,6 +65,18 @@ def my_circuit():
...
+def test_wirecut_adjoint():
+    """Check that the adjoint for the WireCut is working"""
+    dev = qml.device("default.qubit", wires=1)
+
+    @qml.qnode(dev)
+    def my_circuit():
+        adjoint(qml.WireCut)(wires=0)
+        return qml.state()
+
+    assert my_circuit()[0] == 1.0
```
Comment (trbromley) [inline tests/transforms/test_adjoint.py:77]:
```suggestion
    assert np.isclose(my_circuit()[0], 1.0)
```
I get paranoid about equality with floats!

**Review by trbromley — APPROVED** [review]
Thanks @anthayes92, approved! :green_circle:

**Pull request merged**